README.md
date-format
===========

node.js formatting of Date objects as strings. Probably exactly the same as some other library out there.

```sh
npm install date-format
```

usage
=====

Formatting dates as strings
----

```javascript
var format = require('date-format');
format.asString(); //defaults to ISO8601 format and current date.
format.asString(new Date()); //defaults to ISO8601 format
format.asString('hh:mm:ss.SSS', new Date()); //just the time
```

or

```javascript
var format = require('date-format');
format(); //defaults to ISO8601 format and current date.
format(new Date());
format('hh:mm:ss.SSS', new Date());
```

Format string can be anything, but the following letters will be replaced (and leading zeroes added if necessary):
* dd - `date.getDate()`
* MM - `date.getMonth() + 1`
* yy - `date.getFullYear().toString().substring(2, 4)`
* yyyy - `date.getFullYear()`
* hh - `date.getHours()`
* mm - `date.getMinutes()`
* ss - `date.getSeconds()`
* SSS - `date.getMilliseconds()`
* O - timezone offset in ±hhmm format (note that time will still be local if displaying offset)

Built-in formats:
* `format.ISO8601_FORMAT` - `2017-03-14T14:10:20.391` (local time used)
* `format.ISO8601_WITH_TZ_OFFSET_FORMAT` - `2017-03-14T14:10:20.391+11:00` (local + TZ used)
* `format.DATETIME_FORMAT` - `14 03 2017 14:10:20.391` (local time used)
* `format.ABSOLUTETIME_FORMAT` - `14:10:20.391` (local time used)

Parsing strings as dates
----
The date format library has limited ability to parse strings into dates. It can convert strings created using date format patterns (as above), but if you're looking for anything more sophisticated than that you should probably look for a better library ([momentjs](https://momentjs.com) does pretty much everything).

```javascript
var format = require('date-format');
// pass in the format of the string as first argument
format.parse(format.ISO8601_FORMAT, '2017-03-14T14:10:20.391');
// returns Date
```


CHANGELOG.md
# date-format Changelog

## 4.0.13

- [ci: added tests for 4.x, 6.x, 8.x, 10.x, 18.x](https://github.com/nomiddlename/date-format/pull/75) - thanks [@lamweili](https://github.com/lamweili)
  - [chore(deps-dev): bump hosted-git-info from 2.6.0 to 2.8.9](https://github.com/nomiddlename/date-format/pull/76) - thanks [@Dependabot](https://github.com/dependabot)
  - [chore(deps-dev): bump handlebars from 4.0.11 to 4.7.7 ](https://github.com/nomiddlename/date-format/pull/77) - thanks [@Dependabot](https://github.com/dependabot)
  - [chore(deps-dev): bump y18n from 3.2.1 to 3.2.2](https://github.com/nomiddlename/date-format/pull/78) - thanks [@Dependabot](https://github.com/dependabot)

## 4.0.12

- [ci: added tests for 8.x, 10.x, 18.x](https://github.com/nomiddlename/date-format/pull/73) - thanks [@lamweili](https://github.com/lamweili)
- [docs: renamed peteriman to lamweili](https://github.com/nomiddlename/date-format/pull/72) - thanks [@lamweili](https://github.com/lamweili)

## 4.0.11

- [fix: date parsing errors (wrong month due to days overflow)](https://github.com/nomiddlename/date-format/pull/68) - thanks [@lamweili](https://github.com/lamweili)
  - [test: use new Date(0) instead of new Date() before setting every field]() - thanks [@lamweili](https://github.com/lamweili)
- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/70) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): bump eslint from 8.15.0 to 8.16.0
  - chore(deps-dev): bump eslint-plugin-mocha from 10.0.4 to 10.0.5
  - chore(deps-dev): updated package-lock.json

## 4.0.10

- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/66) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): bump eslint from 8.14.0 to 8.15.0
  - chore(deps-dev): bump mocha from 9.2.2 to 10.0.0
  - chore(deps-dev): updated package-lock.json

## 4.0.9

- build: is exactly the same as 4.0.8 and is a re-published 4.0.8 to npm

## 4.0.8

- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/60) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): bump eslint from 8.13.0 to 8.14.0
  - chore(deps-dev): updated package-lock.json
- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/59) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): updated package-lock.json

## 4.0.7

- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/57) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): bump eslint-plugin-mocha from 10.0.3 to 10.0.4
  - chore(deps-dev): updated package-lock.json
- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/54) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): bump eslint from 8.11.0 to 8.13.0
  - chore(deps-dev): updated package-lock.json

## 4.0.6

- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/52) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): updated package-lock.json

## 4.0.5

- [test: better test coverage instead of ignoring](https://github.com/nomiddlename/date-format/pull/48) - thanks [@lamweili](https://github.com/lamweili)
- [docs: updated README.md with badges](https://github.com/nomiddlename/date-format/pull/50) thanks [@lamweili](https://github.com/lamweili)
- [chore(deps-dev): updated dependencies](https://github.com/nomiddlename/date-format/pull/49) - thanks [@lamweili](https://github.com/lamweili)
  - chore(deps-dev): bump eslint from 8.10.0 to 8.11.0
  - chore(deps-dev): bump mocha from 9.2.1 to 9.2.2
  - chore(deps-dev): updated package-lock.json

## 4.0.4

- chore(deps-dev): updated dependencies - thanks [@lamweili](https://github.com/lamweili)
  - [chore(deps-dev): bump eslint from 8.8.0 to 8.10.0 and mocha from 9.2.0 to 9.2.1](https://github.com/nomiddlename/date-format/pull/46) 
  - [chore(deps-dev): bump eslint from 8.7.0 to 8.8.0 and mocha from 9.1.4 to 9.2.0](https://github.com/nomiddlename/date-format/pull/45) 
  - [chore(deps-dev): updated package-lock.json](https://github.com/nomiddlename/date-format/pull/44) 

## 4.0.3

- [test: 100% test coverage](https://github.com/nomiddlename/date-format/pull/42) - thanks [@lamweili](https://github.com/lamweili)
- chore(deps-dev): updated dependencies - thanks [@lamweili](https://github.com/lamweili)
  - [chore(deps-dev): bump eslint from 8.6.0 to 8.7.0 and mocha from 9.1.3 to 9.1.4](https://github.com/nomiddlename/date-format/pull/41) 

## 4.0.2

- [build: not to publish misc files to NPM](https://github.com/nomiddlename/date-format/pull/39) - thanks [@lamweili](https://github.com/lamweili)
- docs: CHANGELOG.md
  - [docs: removed "log4js" from title of CHANGELOG.md](https://github.com/nomiddlename/date-format/pull/37) - thanks [@joshuabremerdexcom](https://github.com/joshuabremerdexcom)
  - [docs: added "date-format" to title of CHANGELOG.md](https://github.com/nomiddlename/date-format/commit/64a95d0386853692d7d65174f94a0751e775f7ce#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed) - thanks [@lamweili](https://github.com/lamweili)
- chore(deps-dev): updated dependencies - thanks [@lamweili](https://github.com/lamweili)
  - [chore(deps-dev): bump eslint-plugin-mocha from 5.3.0 to 10.0.3](https://github.com/nomiddlename/date-format/pull/38) 

## 4.0.1

- build: is exactly the same as 4.0.0 and is a re-published 4.0.0 to npm

## 4.0.0

- [fix: timezone format to include colon separator](https://github.com/nomiddlename/date-format/pull/27) - thanks [@lamweili](https://github.com/lamweili)
  - [test: have a test case for timezone with colon](https://github.com/nomiddlename/date-format/pull/32) - thanks [@lamweili](https://github.com/lamweili)

test/parse-test.js
"use strict";

require("should");
var dateFormat = require("../lib");

describe("dateFormat.parse", function() {
  it("should require a pattern", function() {
    (function() {
      dateFormat.parse();
    }.should.throw(/pattern must be supplied/));
    (function() {
      dateFormat.parse(null);
    }.should.throw(/pattern must be supplied/));
    (function() {
      dateFormat.parse("");
    }.should.throw(/pattern must be supplied/));
  });

  describe("with a pattern that has no replacements", function() {
    it("should return a new date when the string matches", function() {
      dateFormat.parse("cheese", "cheese").should.be.a.Date();
    });

    it("should throw if the string does not match", function() {
      (function() {
        dateFormat.parse("cheese", "biscuits");
      }.should.throw(/String 'biscuits' could not be parsed as 'cheese'/));
    });
  });

  describe("with a full pattern", function() {
    var pattern = "yyyy-MM-dd hh:mm:ss.SSSO";

    it("should return the correct date if the string matches", function() {
      var testDate = new Date(0);
      testDate.setUTCFullYear(2018);
      testDate.setUTCMonth(8);
      testDate.setUTCDate(13);
      testDate.setUTCHours(18);
      testDate.setUTCMinutes(10);
      testDate.setUTCSeconds(12);
      testDate.setUTCMilliseconds(392);

      dateFormat
        .parse(pattern, "2018-09-14 04:10:12.392+1000")
        .getTime()
        .should.eql(testDate.getTime())
        ;
    });

    it("should throw if the string does not match", function() {
      (function() {
        dateFormat.parse(pattern, "biscuits");
      }.should.throw(
        /String 'biscuits' could not be parsed as 'yyyy-MM-dd hh:mm:ss.SSSO'/
      ));
    });
  });

  describe("with a partial pattern", function() {
    var testDate = dateFormat.now() || new Date();
    dateFormat.now = function() {
      return testDate;
    };

    /**
     * If there's no timezone in the format, then we verify against the local date
     */
    function verifyLocalDate(actual, expected) {
      function hasValue(obj) {
        return (obj !== null && obj !== undefined);
      }
      actual.getFullYear().should.eql(hasValue(expected.year) ? expected.year : testDate.getFullYear());
      actual.getMonth().should.eql(hasValue(expected.month) ? expected.month : testDate.getMonth());
      actual.getDate().should.eql(hasValue(expected.day) ? expected.day : testDate.getDate());
      actual.getHours().should.eql(hasValue(expected.hours) ? expected.hours : testDate.getHours());
      actual.getMinutes().should.eql(hasValue(expected.minutes) ? expected.minutes : testDate.getMinutes());
      actual.getSeconds().should.eql(hasValue(expected.seconds) ? expected.seconds : testDate.getSeconds());
      actual
        .getMilliseconds()
        .should.eql(hasValue(expected.milliseconds) ? expected.milliseconds : testDate.getMilliseconds());
    }

    /**
     * If a timezone is specified, let's verify against the UTC time it is supposed to be
     */
    function verifyDate(actual, expected) {
      function hasValue(obj) {
        return (obj !== null && obj !== undefined);
      }
      actual.getUTCFullYear().should.eql(hasValue(expected.year) ? expected.year : testDate.getUTCFullYear());
      actual.getUTCMonth().should.eql(hasValue(expected.month) ? expected.month : testDate.getUTCMonth());
      actual.getUTCDate().should.eql(hasValue(expected.day) ? expected.day : testDate.getUTCDate());
      actual.getUTCHours().should.eql(hasValue(expected.hours) ? expected.hours : testDate.getUTCHours());
      actual.getUTCMinutes().should.eql(hasValue(expected.minutes) ? expected.minutes : testDate.getUTCMinutes());
      actual.getUTCSeconds().should.eql(hasValue(expected.seconds) ? expected.seconds : testDate.getUTCSeconds());
      actual
        .getMilliseconds()
        .should.eql(hasValue(expected.milliseconds) ? expected.milliseconds : testDate.getMilliseconds());
    }

    it("should return a date with missing values defaulting to current time", function() {
      var date = dateFormat.parse("yyyy-MM", "2015-09");
      verifyLocalDate(date, { year: 2015, month: 8 });
    });

    it("should use a passed in date for missing values", function() {
      var missingValueDate = new Date(2010, 1, 8, 22, 30, 12, 100);
      var date = dateFormat.parse("yyyy-MM", "2015-09", missingValueDate);
      verifyLocalDate(date, {
        year: 2015,
        month: 8,
        day: 8,
        hours: 22,
        minutes: 30,
        seconds: 12,
        milliseconds: 100
      });
    });

    it("should handle variations on the same pattern", function() {
      var date = dateFormat.parse("MM-yyyy", "09-2015");
      verifyLocalDate(date, { year: 2015, month: 8 });

      date = dateFormat.parse("yyyy MM", "2015 09");
      verifyLocalDate(date, { year: 2015, month: 8 });

      date = dateFormat.parse("MM, yyyy.", "09, 2015.");
      verifyLocalDate(date, { year: 2015, month: 8 });
    });

    describe("should match all the date parts", function() {
      it("works with dd", function() {
        var date = dateFormat.parse("dd", "21");
        verifyLocalDate(date, { day: 21 });
      });

      it("works with hh", function() {
        var date = dateFormat.parse("hh", "12");
        verifyLocalDate(date, { hours: 12 });
      });

      it("works with mm", function() {
        var date = dateFormat.parse("mm", "34");
        verifyLocalDate(date, { minutes: 34 });
      });

      it("works with ss", function() {
        var date = dateFormat.parse("ss", "59");
        verifyLocalDate(date, { seconds: 59 });
      });

      it("works with ss.SSS", function() {
        var date = dateFormat.parse("ss.SSS", "23.452");
        verifyLocalDate(date, { seconds: 23, milliseconds: 452 });
      });

      it("works with hh:mm O (+1000)", function() {
        var date = dateFormat.parse("hh:mm O", "05:23 +1000");
        verifyDate(date, { hours: 19, minutes: 23 });
      });

      it("works with hh:mm O (-200)", function() {
        var date = dateFormat.parse("hh:mm O", "05:23 -200");
        verifyDate(date, { hours: 7, minutes: 23 });
      });

      it("works with hh:mm O (+09:30)", function() {
        var date = dateFormat.parse("hh:mm O", "05:23 +09:30");
        verifyDate(date, { hours: 19, minutes: 53 });
      });

      it("works with hh:mm O (Z)", function() {
        var date = dateFormat.parse("hh:mm O", "05:23 Z");
        verifyDate(date, { hours: 5, minutes: 23 });
      });
    });
  });

  describe("with a partial pattern and last day of month", function() {
    var testDate = new Date("2022-05-31T00:00:00.000Z");

    /**
     * If there's no timezone in the format, then we verify against the local date
     */
    function verifyLocalDate(actual, expected) {
      function hasValue(obj) {
        return (obj !== null && obj !== undefined);
      }
      actual.getFullYear().should.eql(hasValue(expected.year) ? expected.year : testDate.getFullYear());
      actual.getMonth().should.eql(hasValue(expected.month) ? expected.month : testDate.getMonth());
      actual.getDate().should.eql(hasValue(expected.day) ? expected.day : testDate.getDate());
      actual.getHours().should.eql(hasValue(expected.hours) ? expected.hours : testDate.getHours());
      actual.getMinutes().should.eql(hasValue(expected.minutes) ? expected.minutes : testDate.getMinutes());
      actual.getSeconds().should.eql(hasValue(expected.seconds) ? expected.seconds : testDate.getSeconds());
      actual
        .getMilliseconds()
        .should.eql(hasValue(expected.milliseconds) ? expected.milliseconds : testDate.getMilliseconds());
    }

    it("should return a date with missing values defaulting to current time", function() {
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-01-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 0, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-02-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 1, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-03-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 2, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-04-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 3, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-05-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 4, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-06-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 5, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-07-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 6, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-08-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 7, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-09-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 8, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-10-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 9, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-11-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 10, day: 1});
      verifyLocalDate(dateFormat.parse('yyyy-MM-dd', '2022-12-01', new Date("2022-05-31T00:00:00.000Z")), { year: 2022, month: 11, day: 1});
    });
  });

  describe("with a date formatted by this library", function() {
    describe("should format and then parse back to the same date", function() {
      function testDateInitWithUTC() {
        var td = new Date(0);
        td.setUTCFullYear(2018);
        td.setUTCMonth(8);
        td.setUTCDate(13);
        td.setUTCHours(18);
        td.setUTCMinutes(10);
        td.setUTCSeconds(12);
        td.setUTCMilliseconds(392);
        return td;
      }

      it("works with ISO8601_WITH_TZ_OFFSET_FORMAT", function() {
        // For this test case to work, the date object must be initialized with
        // UTC timezone
        var td = testDateInitWithUTC();
        var d = dateFormat(dateFormat.ISO8601_WITH_TZ_OFFSET_FORMAT, td);
        dateFormat.parse(dateFormat.ISO8601_WITH_TZ_OFFSET_FORMAT, d)
          .should.eql(td);
      });

      it("works with ISO8601_FORMAT", function() {
        var td = new Date();
        var d = dateFormat(dateFormat.ISO8601_FORMAT, td);
        var actual = dateFormat.parse(dateFormat.ISO8601_FORMAT, d);
        actual.should.eql(td);
      });

      it("works with DATETIME_FORMAT", function() {
        var testDate = new Date();
        dateFormat
        .parse(
          dateFormat.DATETIME_FORMAT,
          dateFormat(dateFormat.DATETIME_FORMAT, testDate)
        )
        .should.eql(testDate);
      });

      it("works with ABSOLUTETIME_FORMAT", function() {
        var testDate = new Date();
        dateFormat
        .parse(
          dateFormat.ABSOLUTETIME_FORMAT,
          dateFormat(dateFormat.ABSOLUTETIME_FORMAT, testDate)
        )
        .should.eql(testDate);
      });
    });
  });
});


test/date-format_test.js
"use strict";
var should = require('should')
, dateFormat = require('../');

describe('date-format', function() {
  var date = new Date(2010, 0, 11, 14, 31, 30, 5);

  it('should format a date as string using a pattern', function() {
    dateFormat.asString(dateFormat.DATETIME_FORMAT, date).should.eql("11 01 2010 14:31:30.005");
  });

  it('should default to the ISO8601 format', function() {
    dateFormat.asString(date).should.eql('2010-01-11 14:31:30.005');
  });

  it('should provide a ISO8601 with timezone offset format', function() {
    date.getTimezoneOffset = function() { return -660; };
    dateFormat.asString(
      dateFormat.ISO8601_WITH_TZ_OFFSET_FORMAT, 
      date
    ).should.eql(
      "2010-01-11T14:31:30+1100"
    );

    date.getTimezoneOffset = function() { return 120; };
    dateFormat.asString(
      dateFormat.ISO8601_WITH_TZ_OFFSET_FORMAT, 
      date
    ).should.eql(
      "2010-01-11T14:31:30-0200"
    );
  });

  it('should provide a just-the-time format', function() {
    dateFormat.asString(dateFormat.ABSOLUTETIME_FORMAT, date).should.eql('14:31:30.005');
  });

  it('should provide a custom format', function() {
    date.getTimezoneOffset = function() { return 120; };
    dateFormat.asString("O.SSS.ss.mm.hh.dd.MM.yy", date).should.eql('-0200.005.30.31.14.11.01.10');
  });

  it('should format a date as string using name of day and month in english', function() {
    var monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'Jully', 'August', 'September', 'October', 'November', 'December'],
        dayNames = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    dateFormat("MMMM the ddth, yyyy. It's a DDDD.", date, monthNames, dayNames).should.eql("January the 11th, 2010. It's a Monday.");
  });

  it('should format a date as string using name of day and month in french', function() {
    var monthNames = ['Janvier', 'Février', 'Mars', 'Avril', 'Mai', 'Juin', 'Juillet', 'Août', 'Septembre', 'Octobre', 'Novembre', 'Decembre'],
        dayNames = ['Dimanche', 'Lundi', 'Mardi', 'Mercredi', 'Jeudi', 'Vendredi', 'Samedi'];
    dateFormat("Le DDDD dd MMMM yyyy à hh:mm:ss", date, monthNames, dayNames).should.eql("Le Lundi 11 Janvier 2010 à 14:31:30");
  });

  it('should format a date as string using name of day and month without array', function() {
    dateFormat("It's day number DDDD and month number MMMM.", date).should.eql("It's day number 2 and month number 1.");
  });

  it('should format a date as string using name of day and month without array', function() {
    dateFormat("It's day number DD and month number MM.", date).should.eql("It's day number 02 and month number 01.");
  });

  it('should format a date with text', function() {
    dateFormat("Community Address Orbit vs Co\\mmunity A\\ddre\\ss \\Orbit.", date).should.eql("Co31unity A11re30 -0200rbit vs Community Address Orbit.");
  });
});


test/date_format-test.js
'use strict';

require('should');

var dateFormat = require('../lib');

function createFixedDate() {
    return new Date(2010, 0, 11, 14, 31, 30, 5);
}

describe('date_format', function() {
    var date = createFixedDate();

    it('should default to now when a date is not provided', function() {
        dateFormat.asString(dateFormat.DATETIME_FORMAT).should.not.be.empty();
    });

    it('should be usable directly without calling asString', function() {
        dateFormat(dateFormat.DATETIME_FORMAT, date).should.eql('11 01 2010 14:31:30.005');
    });

    it('should format a date as string using a pattern', function() {
        dateFormat.asString(dateFormat.DATETIME_FORMAT, date).should.eql('11 01 2010 14:31:30.005');
    });

    it('should default to the ISO8601 format', function() {
        dateFormat.asString(date).should.eql('2010-01-11T14:31:30.005');
    });

    it('should provide a ISO8601 with timezone offset format', function() {
        var tzDate = createFixedDate();
        tzDate.getTimezoneOffset = function () {
            return -660;
        };

        // when tz offset is in the pattern, the date should be in local time
        dateFormat.asString(dateFormat.ISO8601_WITH_TZ_OFFSET_FORMAT, tzDate)
            .should.eql('2010-01-11T14:31:30.005+11:00');

        tzDate = createFixedDate();
        tzDate.getTimezoneOffset = function () {
            return 120;
        };

        dateFormat.asString(dateFormat.ISO8601_WITH_TZ_OFFSET_FORMAT, tzDate)
            .should.eql('2010-01-11T14:31:30.005-02:00');

        tzDate = createFixedDate();
        tzDate.getTimezoneOffset = function () {
            return 0;
        };

        dateFormat.asString(dateFormat.ISO8601_WITH_TZ_OFFSET_FORMAT, tzDate)
            .should.eql('2010-01-11T14:31:30.005Z');
    });

    it('should provide a just-the-time format', function() {
        dateFormat.asString(dateFormat.ABSOLUTETIME_FORMAT, date).should.eql('14:31:30.005');
    });

    it('should provide a custom format', function() {
        var customDate = createFixedDate();
        customDate.getTimezoneOffset = function () {
            return 120;
        };

        dateFormat.asString('O.SSS.ss.mm.hh.dd.MM.yy', customDate).should.eql('-02:00.005.30.31.14.11.01.10');
    });
});


lib/index.js
"use strict";

function padWithZeros(vNumber, width) {
  var numAsString = vNumber.toString();
  while (numAsString.length < width) {
    numAsString = "0" + numAsString;
  }
  return numAsString;
}

function addZero(vNumber) {
  return padWithZeros(vNumber, 2);
}

/**
 * Formats the TimeOffset
 * Thanks to http://www.svendtofte.com/code/date_format/
 * @private
 */
function offset(timezoneOffset) {
  var os = Math.abs(timezoneOffset);
  var h = String(Math.floor(os / 60));
  var m = String(os % 60);
  if (h.length === 1) {
    h = "0" + h;
  }
  if (m.length === 1) {
    m = "0" + m;
  }
  return timezoneOffset === 0 ? "Z" : (timezoneOffset < 0 ? "+" : "-") + h + ":" + m;
}

function asString(format, date) {
  if (typeof format !== "string") {
    date = format;
    format = module.exports.ISO8601_FORMAT;
  }
  if (!date) {
    date = module.exports.now();
  }

  // Issue # 14 - Per ISO8601 standard, the time string should be local time
  // with timezone info.
  // See https://en.wikipedia.org/wiki/ISO_8601 section "Time offsets from UTC"

  var vDay = addZero(date.getDate());
  var vMonth = addZero(date.getMonth() + 1);
  var vYearLong = addZero(date.getFullYear());
  var vYearShort = addZero(vYearLong.substring(2, 4));
  var vYear = format.indexOf("yyyy") > -1 ? vYearLong : vYearShort;
  var vHour = addZero(date.getHours());
  var vMinute = addZero(date.getMinutes());
  var vSecond = addZero(date.getSeconds());
  var vMillisecond = padWithZeros(date.getMilliseconds(), 3);
  var vTimeZone = offset(date.getTimezoneOffset());
  var formatted = format
    .replace(/dd/g, vDay)
    .replace(/MM/g, vMonth)
    .replace(/y{1,4}/g, vYear)
    .replace(/hh/g, vHour)
    .replace(/mm/g, vMinute)
    .replace(/ss/g, vSecond)
    .replace(/SSS/g, vMillisecond)
    .replace(/O/g, vTimeZone);
  return formatted;
}

function setDatePart(date, part, value, local) {
  date['set' + (local ? '' : 'UTC') + part](value);
}

function extractDateParts(pattern, str, missingValuesDate) {
  // Javascript Date object doesn't support custom timezone.  Sets all felds as
  // GMT based to begin with.  If the timezone offset is provided, then adjust
  // it using provided timezone, otherwise, adjust it with the system timezone.
  var local = pattern.indexOf('O') < 0;
  var matchers = [
    {
      pattern: /y{1,4}/,
      regexp: "\\d{1,4}",
      fn: function(date, value) {
        setDatePart(date, 'FullYear', value, local);
      }
    },
    {
      pattern: /MM/,
      regexp: "\\d{1,2}",
      fn: function(date, value) {
        setDatePart(date, 'Month', (value - 1), local);
      }
    },
    {
      pattern: /dd/,
      regexp: "\\d{1,2}",
      fn: function(date, value) {
        setDatePart(date, 'Date', value, local);
      }
    },
    {
      pattern: /hh/,
      regexp: "\\d{1,2}",
      fn: function(date, value) {
        setDatePart(date, 'Hours', value, local);
      }
    },
    {
      pattern: /mm/,
      regexp: "\\d\\d",
      fn: function(date, value) {
        setDatePart(date, 'Minutes', value, local);
      }
    },
    {
      pattern: /ss/,
      regexp: "\\d\\d",
      fn: function(date, value) {
        setDatePart(date, 'Seconds', value, local);
      }
    },
    {
      pattern: /SSS/,
      regexp: "\\d\\d\\d",
      fn: function(date, value) {
        setDatePart(date, 'Milliseconds', value, local);
      }
    },
    {
      pattern: /O/,
      regexp: "[+-]\\d{1,2}:?\\d{2}?|Z",
      fn: function(date, value) {
        if (value === "Z") {
          value = 0;
        }
        else {
          value = value.replace(":", "");
        }
        var offset = Math.abs(value);
        var timezoneOffset = (value > 0 ? -1 :  1 ) * ((offset % 100) + Math.floor(offset / 100) * 60);
        // Per ISO8601 standard: UTC = local time - offset
        //
        // For example, 2000-01-01T01:00:00-0700
        //   local time: 2000-01-01T01:00:00
        //   ==> UTC   : 2000-01-01T08:00:00 ( 01 - (-7) = 8 )
        //
        // To make it even more confusing, the date.getTimezoneOffset() is
        // opposite sign of offset string in the ISO8601 standard.  So if offset
        // is '-0700' the getTimezoneOffset() would be (+)420. The line above
        // calculates timezoneOffset to matche Javascript's behavior.
        //
        // The date/time of the input is actually the local time, so the date
        // object that was constructed is actually local time even thought the
        // UTC setters are used.  This means the date object's internal UTC
        // representation was wrong.  It needs to be fixed by substracting the
        // offset (or adding the offset minutes as they are opposite sign).
        //
        // Note: the time zone has to be processed after all other fileds are
        // set.  The result would be incorrect if the offset was calculated
        // first then overriden by the other filed setters.
        date.setUTCMinutes(date.getUTCMinutes() + timezoneOffset);
      }
    }
  ];

  var parsedPattern = matchers.reduce(
    function(p, m) {
      if (m.pattern.test(p.regexp)) {
        m.index = p.regexp.match(m.pattern).index;
        p.regexp = p.regexp.replace(m.pattern, "(" + m.regexp + ")");
      } else {
        m.index = -1;
      }
      return p;
    },
    { regexp: pattern, index: [] }
  );

  var dateFns = matchers.filter(function(m) {
    return m.index > -1;
  });
  dateFns.sort(function(a, b) {
    return a.index - b.index;
  });

  var matcher = new RegExp(parsedPattern.regexp);
  var matches = matcher.exec(str);
  if (matches) {
    var date = missingValuesDate || module.exports.now();
    dateFns.forEach(function(f, i) {
      f.fn(date, matches[i + 1]);
    });

    return date;
  }

  throw new Error(
    "String '" + str + "' could not be parsed as '" + pattern + "'"
  );
}

function parse(pattern, str, missingValuesDate) {
  if (!pattern) {
    throw new Error("pattern must be supplied");
  }

  return extractDateParts(pattern, str, missingValuesDate);
}

/**
 * Used for testing - replace this function with a fixed date.
 */
function now() {
  return new Date();
}

module.exports = asString;
module.exports.asString = asString;
module.exports.parse = parse;
module.exports.now = now;
module.exports.ISO8601_FORMAT = "yyyy-MM-ddThh:mm:ss.SSS";
module.exports.ISO8601_WITH_TZ_OFFSET_FORMAT = "yyyy-MM-ddThh:mm:ss.SSSO";
module.exports.DATETIME_FORMAT = "dd MM yyyy hh:mm:ss.SSS";
module.exports.ABSOLUTETIME_FORMAT = "hh:mm:ss.SSS";
